Piped communication on Windows

TGWDB opened the pull request:
Piped inter-process communication for Windows.

**Old text below.**

This is a PR to test CI and see how the piped communication progresses on Windows. This started based on #6151 and so will need to either wait for that to be merged and rebased, or supersede that PR.

Note that this is part of the larger SMT plan available at #6134 .

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,361 @@
+  sec_attr.nLength = sizeof(SECURITY_ATTRIBUTES);
+  sec_attr.bInheritHandle = TRUE;
+  sec_attr.lpSecurityDescriptor = NULL;
+  // Use named pipes to allow non-blocking read
```
What access rights does this security descriptor grant? The default one for named pipes gives read access to the EVERYONE group.

TGWDB replied:
If I read the documentation [here](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/aa379560(v=vs.85)) correctly then by defining the security attributes for `null` for `lpSecurityDescriptor` (`sec_attr` approx line 30 of `src/util/piped_process.cpp`) this sets the security to be the default for the current session access token. In particular this is NOT available to all users.

That said, I'll add comment about this to the code.

TGWDB replied:
Added comment to code as well.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,361 @@
+  // Use named pipes to allow non-blocking read
+  child_std_IN_Rd = CreateNamedPipe(
+      TEXT("\\\\.\\pipe\\cbmc\\SMT2\\child\\IN"),
+      PIPE_ACCESS_INBOUND,          // Reading for us
```
What happens when there are multiple instances of CBMC running simultaneously? Do they all try to talk over the same named pipe?

TGWDB replied:
Fixed with use of random number.

kroening replied:
`rand` isn't random, but produces a deterministic sequence unless you change the seed.
Consider using something you know to be unique, say a process ID.

TGWDB replied:
It was seeded, but process ID is better - done.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,361 @@
+  for(int i = 0; i < commandvec.size(); i++)
+  {
+    cmdline.append(commandvec[i].c_str());
+    cmdline.append(" ");
```
This will require quoting. Use `quote_windows_arg` from util/run.cpp

TGWDB replied:
I'd prefer to leave this to the calling code to properly ensure the handling of quotations (also this breaks unit tests and makes for messy dependencies for now).

kroening replied:
This is a no-go, since it has security implications.

TGWDB replied:
Added quoting here.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,361 @@
+    }
+// TODO, make this define and choice better
+#  define WIN_POLL_WAIT 10
+    Sleep(WIN_POLL_WAIT);
```
There is no better option than polling? This will heat a lot of CPUs!

TGWDB replied:
I haven't found evidence of one that works nicely with our asynchronous reading from named pipes. The options available for waiting appear to work with blocking pipes, but this does not work with our interactive approach (at least at the moment).

If someone who knows Windows better can suggest an alternative I'll look into it.

kroening replied:
I'd say this is a blocker for using this.

kroening replied:
`MsgWaitForMultipleObjectsEx` might do the right thing.

TGWDB replied:
`WaitForMultipleObjects` (and similar ones) rely upon the events they wait on being synchronous. This (so far in testing and from what I've read) means we have to wait for the child process to finish before we can read from the pipe. I saw some example code that used this approach with asynchronous pipes, but then used a busy loop with `peek` in the same way I've implemented currently.

I'll keep looking, and play around more now I've sorted out most of the other quirks.

TGWDB replied:
See comment/discussion at the top of the file for more on why this is done this way.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,383 @@
+  TCHAR *szCmdline =
+    reinterpret_cast<TCHAR *>(malloc(sizeof(TCHAR) * cmdline.size()));
+  _tcscpy(szCmdline, cmdline.c_str());
+  success = CreateProcess(
```
This should use `CreateProcessW`, and `widen(cmdline)`

TGWDB replied:
Done, along with quoting.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,383 @@
+  TerminateProcess(proc_info.hProcess, 0);
+  CloseHandle(proc_info.hProcess);
+  CloseHandle(proc_info.hThread);
+#else
```
Does this clean up the named pipe?

TGWDB replied:
According to the docs when all handles are closed/released then Windows cleans up the pipe. I've pushed some more aggressive killing and disconnecting of the child process, but it probably makes no difference in practice.

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,388 @@
+  // and causes heap corruption in Windows if we free!
+  WCHAR *szCmdline =
+    reinterpret_cast<WCHAR *>(malloc(sizeof(WCHAR) * cmdline.size()));
+  wcscpy(szCmdline, cmdline.c_str());
```
You need to add one `WCHAR` to the size of the allocated buffer for the terminating zero.
To avoid this problem altogether, use `_wcsdup`.

TGWDB replied:
Done

kroening commented on src/util/piped_process.cpp:
```diff
@@ -0,0 +1,388 @@
+    while(i < commandvec.size())
+    {
+      args[i] = strdup(commandvec[i].c_str());
+      i++;
```
No need for the `strdup`, and the subsequent `free`.

TGWDB replied:
The child process wants to be able to modify the arguments (i.e. non `const`) while the `commandvec`  with `c_str()`can only provide `const` instances.

kroening replied:
This memory is not shared with the child; the child gets a copy. `const` is fine.

TGWDB replied:
Fixed. :)

kroening replied:
Please add the `const` to the type of `args`. There is then no need for the `const_cast`.

TGWDB committed:
Add in Windows version of piped_process functions

This adds in windows versions of the piped_process communication
functions.

TGWDB committed:
Expose quote_windows_arg

Make the quote_windows_arg function available for use

TGWDB committed:
Add unit tests for Windows piped process IPC

Adds unit tests for Windows piped process communication. These duplicate
the Linux/MacOS ones in most cases. There is also a termination one that
does not have a working Linux/MacOS version. (Also z3 is not tested on
Windows).

TGWDB commented:
Now rebased on develop after merging of #6151 .

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+  }
+  // Note that we do NOT free this since it becomes part of the child
+  // and causes heap corruption in Windows if we free!
+  success = CreateProcessW(
```
:pick: I suggest declaring success where it is used - `const BOOL success = CreateProcessW(`

TGWDB replied:
Fixed

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+  start_info.hStdInput = child_std_IN_Rd;
+  start_info.dwFlags |= STARTF_USESTDHANDLES;
+  // Unpack the command into a single string for Windows API
+  std::wstring cmdline = widen(commandvec[0]);
```
:pick: I suggest splitting the lines of code which build `cmdline` out into a separate function and then declaring the variable holding the string returned as `const`.

TGWDB replied:
Fixed

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -128,15 +326,17 @@ piped_processt::~piped_processt()
   if(process_state != statet::CREATED)
   {
-    return send_responset::ERROR;
+    return send_responset::ERRORED;
```
:pick: I would have kept this rename in a separate commit, in order to keep the commits smaller and restricted to a single change. I consider this something to consider for next time, rather than something worth editing the commit history to fix.

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -107,15 +295,25 @@ piped_processt::piped_processt(const std::vector<std::string> commandvec)
 {
 #  ifdef _WIN32
-  UNIMPLEMENTED_FEATURE("Pipe IPC on windows: piped_processt constructor")
+  TerminateProcess(proc_info.hProcess, 0);
```
:no_entry_sign: :question: Are we violating the C++ rule of three which says that if we define a custom destructor then we also need to define the copy constructor and copy assignment operator? I wasn't sure if these other operations were implicitly deleted. If we violate this rule then we are at risk of copying this class and then freeing the resources it handles multiple times. The simplest fix for this issue could be to explicitly delete the other 2, to prevent copying. Alternatively the code in the destructor could be moved into custom deleters for the process info and handles using `std::unique_ptr` / `std::shared_ptr` pointers pointing to the managed resources.

TGWDB replied:
Fixed

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+  // Use process ID as a unique ID for this process at this time.
+  base_name.append(std::to_string(GetCurrentProcessId()));
+  std::string tmp_name = base_name;
+  tmp_name.append("\\IN");
```
:bulb: I suggest using `const std::string in_name = base_name + "\\IN";` rather than using `.append`. This allows us to both make the variable const and use a slightly more succinct syntax.

TGWDB replied:
Fixed

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -145,48 +345,40 @@ piped_processt::send_responset piped_processt::send(const std::string &message)
-      return response;
-    case 0:
-      // Pipe is closed.
-      process_state = statet::STOPPED;
```
:no_entry_sign: :question: Why is the usage of the STOPPED state in the non-Windows version being removed as part of the implementation of the Windows version?

TGWDB replied:
Old behaviour preserved, added comment about to check if we really need this state (TODO)

TGWDB replied:
Removed excess states (in recent separate commit).

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -204,12 +396,28 @@ piped_processt::statet piped_processt::get_status()
+  LPDWORD nbytes = 0;
+  LPDWORD rbytes = 0;
+  LPDWORD rmbytes = 0;
+  while(timeout < 0 || waited_time >= timeout)
```
:pick: The magic value of `-1` for `timeout` is set due to the unix API being used. I suggest the "unwrap" is moved into the non-Windows code and this line be written as `while(!wait_time || waited_time >= *wait_time)`

thomasspriggs commented on src/util/piped_process.h:
```diff
@@ -6,11 +6,10 @@
 #ifdef _WIN32
-// TODO: Windows definitions go here
-#else
+#  include <windows.h>
```
:no_entry_sign: Including `windows.h` in this header file exposes the entire windows API to any other file which includes this one when building on Windows. This has the potential to make it more difficult for contributors working on Windows to write platform independent code and could also increase build times. This issue could be worked around by forward declaring `PROCESS_INFORMATION` and `HANDLE` and making the member fields pointers to these types, or by using the "Pointer to implementation" / "pImpl" technique to only declare fields of these types in the `.cpp` file.

TGWDB replied:
Fixed.

thomasspriggs commented on unit/util/piped_process.cpp:
```diff
@@ -61,6 +66,39 @@ TEST_CASE(
+#ifdef _WIN32
+  commands.push_back("cmd /c ping 127.0.0.1 -n 6 > nul");
+  SYSTEMTIME st;
+  GetSystemTime(&st);
```
:pick: This should probably be done using C++11's `std::chrono::steady_clock` from the `<chrono>` header, so as to avoid unnecessarily making the test operating system specific.

TGWDB replied:
Fixed

thomasspriggs reviewed (changes requested):
Great to see some more progress on this, but there are some :no_entry_sign: issues which we should address before merging this.

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -2,19 +2,90 @@
+// access from the same session token/permissions. This SHOULD be sufficient
+// for what we need.
+//
+// Non-blocking pipes allow immediate reading of any data on the pipre which
```
pipe

TGWDB replied:
Fixed

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -2,19 +2,90 @@
+// for what we need.
+//
+// Non-blocking pipes allow immediate reading of any data on the pipre which
+// matches the Linux/MasOC pipe behvariour and also allows reading of the
```
behaviour

tautschnig replied:
Also: MacOS

TGWDB replied:
Both fixed

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -2,19 +2,90 @@
+//    finished, again undoing the interactive behaviour desired.
+// Since none of the above work effectivley, the chosen approach is to use a
+// non-blocking peek to see if there is anthing to read, and use a sleep and
+// poll behaviour that might be puch busier than we want. At the time of
```
much?

TGWDB replied:
Fixed

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -2,19 +2,90 @@
+//    means the example above cannot read "The Jabberwocky" until the ping has
+//    finished, again undoing the interactive behaviour desired.
+// Since none of the above work effectivley, the chosen approach is to use a
+// non-blocking peek to see if there is anthing to read, and use a sleep and
```
Would something like this work instead of sleep or is this no better but just with more overhead?
```
do {
  // check whether there's something new
} while (WaitForSingleObject(hEvent, 100) == WAIT_TIMEOUT)
```

TGWDB replied:
Those don't work on non-blocking pipes, they immediately return and so would not actually sleep.

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
-  UNIMPLEMENTED_FEATURE("Pipe IPC on windows.")
+  // Security attributes for pipe creation
+  SECURITY_ATTRIBUTES sec_attr;
+  // Ensure pipes are inherited
```
comment should be one line further down

TGWDB replied:
Fixed

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+    PIPE_UNLIMITED_INSTANCES,     // Probably doesn't matter
+    BUFSIZE,
+    BUFSIZE,    // Output and input bufffer sizes
+    0,          // Timeout in ms, 0 = use system default
```
This is the timeout to detect that the other side hasn't connected, I guess?

peterschrammel replied:
Any insights on this? Would be good to explain to save follow-ups the time for investigating how the API works.

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+    &sec_attr); // For inheritance by child
+  if(child_std_IN_Rd == INVALID_HANDLE_VALUE)
+  {
+    throw std::runtime_error("Input pipe creation failed for child_std_IN_Rd");
```
I think we have exception types in `util/exception_util.h` that are more suitable.

TGWDB replied:
Fixed

peterschrammel commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
 
 piped_processt::piped_processt(const std::vector<std::string> commandvec)
 {
+  // Default state
```
Would it be possible to split up this mega-function a bit?

TGWDB replied:
Split out a small helper function, but the function is pretty straightforward, just many lines due to many arguments to function calls to create Windows processes/pipes (also lots of commends/explanation).

tautschnig commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
 
 piped_processt::piped_processt(const std::vector<std::string> commandvec)
```
I recognise this was here before, but any insight as to why this isn't being passed as a reference? The `const` doesn't seem terribly useful at present.

TGWDB replied:
No insight, but fixed now.

tautschnig commented on src/util/piped_process.cpp:
```diff
@@ -30,8 +101,127 @@
+  sec_attr.lpSecurityDescriptor = NULL;
+  // Use named pipes to allow non-blocking read
+  // Build the base name for the pipes
+  std::string base_name = "\\\\.\\pipe\\cbmc\\SMT2\\child\\";
```
This is generic code in `util`, I don't think it should hard-code `SMT2`.

TGWDB replied:
Fixed

TGWDB committed:
Fix missing copy operations (delete them)

This removes the copy operations from the piped_process class
since we do not want to implement these.
Also fix unit tests that used these.

TGWDB committed:
Fix unit tests to use std::chrono not Windows.h

This fixes the unit tests to use the std::chrono library insead
of windows.h for timing information.

TGWDB committed:
Fix destructor scope to fix termination test

Instead of calling the destructor explicitly, use braces to scope
and use implicit destructor calling.

TGWDB committed:
Fix include scoping and use unique_ptr

This fixes the including of windows.h to limited scope. Requires
some forward definition and uses unique_ptr.

TGWDB committed:
Fix to use system_exceptiont

Replace std::* exceptions in Windows code with system_exceptiont

TGWDB committed:
Split out helper function for Windows string construction

Split out a helper function to process an array of strings into
a single wide string for use in Windows commands.

TGWDB committed:
Reduce unecessary states for piped process statet

Reduce the number of possible states for the piped_process statet
since we only really need to know if it's running or has an error
for now.

TGWDB committed:
Pass commandvec by reference and small fixes

This changes to pass the vector of strings for the piped_process
command by reference instead of by value.

Also add some doxygen and static keyword

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -101,6 +101,19 @@
 #  define BUFSIZE 2048
 
+#ifdef _WIN32
+std::wstring process_windows_args(const std::vector<std::string> commandvec)
```
:bulb: I would declare this as `static std::wstring process_windows_args(const std::vector<std::string> commandvec)`. Adding the `static` tells the compiler that this function has internal linkage only. Meaning that it can only be called from within this `.cpp` file. This informs the reader that this function is an internal implementation detail, not part of the external interface.

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -193,13 +206,7 @@ piped_processt::piped_processt(const std::vector<std::string> commandvec)
   start_info.hStdOutput = child_std_OUT_Wr;
   start_info.hStdInput = child_std_IN_Rd;
   start_info.dwFlags |= STARTF_USESTDHANDLES;
-  // Unpack the command into a single string for Windows API
```
:pick: I suggest reusing this comment as doxygen just above the definition of the new function.

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -193,13 +206,7 @@ piped_processt::piped_processt(const std::vector<std::string> commandvec)
-    cmdline.append(L" ");
-    cmdline.append(quote_windows_arg(widen(commandvec[i])));
-  }
+  const std::wstring cmdline = process_windows_args(commandvec);
```
:pick: `process` is a little non-descriptive. Maybe `prepare_windows_command_line` would be better.

thomasspriggs commented on src/util/piped_process.cpp:
```diff
@@ -102,19 +102,19 @@
 
 #ifdef _WIN32
-std::wstring process_windows_args(const std::vector<std::string> commandvec)
+std::wstring process_windows_args(const std::vector<std::string> *commandvec)
```
:no_entry_sign: Adding that `*` passes by pointer, not by reference. To pass by reference you should use `&`. Passing by pointer is an inferior solution as it allows `nullptr` to be passed as an argument, whereas a reference is non-nullable (as is passing by value). This issue should be fixed before this PR is merged, because it affects the way the constructor is used. So if we leave resolving it later then we may need to fix additional usages of the constructor built on top of this PR.

thomasspriggs replied:
Looks to be fixed.

thomasspriggs reviewed (approved):
:+1:

peterschrammel commented on unit/util/piped_process.cpp:
```diff
@@ -35,10 +37,15 @@ TEST_CASE(
-  const std::string expected_error("Launching abcde failed");
   std::vector<std::string> commands;
+#ifdef _WIN32
+  const std::string expected_error("'abcde' is not recogni");
```
```suggestion
  const std::string expected_error("'abcde' is not recognized");
```

peterschrammel commented on unit/util/piped_process.cpp:
```diff
@@ -61,14 +68,51 @@ TEST_CASE(
+      end_time - start_time);
+  size_t calc = time_span.count();
+#else
+  // Currently not working under Linxu/MacOS?!
```
```suggestion
  // Currently not working under Linux/MacOS?!
```

peterschrammel commented on unit/util/piped_process.cpp:
```diff
@@ -61,14 +68,51 @@ TEST_CASE(
+  size_t calc = time_span.count();
+#else
+  // Currently not working under Linxu/MacOS?!
+  // commands.push_back("sleep 6");
```
Put in `#if 0` rather than commenting.

TGWDB committed:
Update comment on timeout in windows

Add comment on timeout in windows.

TGWDB committed:
Tidy up unit test code

Tidy up unit test code to have cleaner string in windows failed
command test, and useful #if for non-working code.
